m2cgen/interpreters/f_sharp/interpreter.py

```diff
@@ -0,0 +1,120 @@
+
+        with self._cg.function_definition(name=func_name, args=args):
+            last_result = self._do_interpret(expr)
+            self._dump_cache()
+            self._cg.add_code_line(last_result)
+
+        if self.with_linear_algebra:
+            filename = os.path.join(
+                os.path.dirname(__file__), "linear_algebra.fs")
+            self._cg.prepend_code_lines(utils.get_file_content(filename))
+
+        if self.with_log1p_expr:
+            filename = os.path.join(
+                os.path.dirname(__file__), "log1p.fs")
+            self._cg.prepend_code_lines(utils.get_file_content(filename))
+
+        return self._cg.finalize_and_get_generated_code()
+
+    def interpret_if_expr(self, expr, if_code_gen=None, **kwargs):
+        if if_code_gen is None:
+            code_gen = FSharpCodeGenerator(indent=self.indent)
+            nested = False
+        else:
+            code_gen = if_code_gen
+            nested = True
+
+        code_gen.add_if_statement(self._do_interpret(
+            expr.test, **kwargs))
+        code_gen.add_code_line(self._do_interpret(
+            expr.body, if_code_gen=code_gen, **kwargs))
+        code_gen.add_else_statement()
+        code_gen.add_code_line(self._do_interpret(
+            expr.orelse, if_code_gen=code_gen, **kwargs))
+        code_gen.add_if_termination()
+
+        if not nested:
+            return self._cache_reused_expr(
+                expr, code_gen.finalize_and_get_generated_code())
+
+    def interpret_pow_expr(self, expr, **kwargs):
+        base_result = self._do_interpret(expr.base_expr, **kwargs)
+        exp_result = self._do_interpret(expr.exp_expr, **kwargs)
+        return self._cg.infix_expression(
+            left=base_result, right=exp_result, op="**")
+
+    def interpret_log1p_expr(self, expr, **kwargs):
+        self.with_log1p_expr = True
+        return super().interpret_log1p_expr(expr, **kwargs)
+
+    # Cached expressions become functions with no arguments, i.e. values
+    # which are CAFs. Therefore, they are computed only once.
+    def _cache_reused_expr(self, expr, expr_result):
+        if expr in self._cached_expr_results:
+            return self._cached_expr_results[expr].var_name
+        else:
+            func_name = self._cg.get_func_name()
+            self._cached_expr_results[expr] = utils.CachedResult(
+                var_name=func_name, expr_result=expr_result)
+            return func_name
+
+    def _dump_cache(self):
+        if self._cached_expr_results:
+            for func_name, expr_result in self._cached_expr_results.values():
+                self._cg.add_function(
+                    function_name=func_name, function_body=expr_result)
+
+    def bin_depth_threshold_hook(self, expr, **kwargs):
+        if expr in self._cached_expr_results:
+            return self._cached_expr_results[expr].var_name
+        result = self._do_interpret(expr, **kwargs)
+        return self._cache_reused_expr(expr, result)
```

Comment:
I think it is OK to use caching in a particular interpreter but not in a base class. Refer to https://github.com/BayesWitnesses/m2cgen/pull/231#discussion_r442295716.